File: src/annealage_mesh/static/js/ws.js
/**
 * Browser side of the /ws transport: reads the per-run token out of the URL
 * once, connects, exchanges hello and replay, dispatches server events, and
 * drives the topbar connection indicator. Reconnection uses capped
 * exponential backoff and never stops on its own except for the two cases
 * that retrying cannot fix within this page load: a protocol-version
 * mismatch, and a confirmed token refusal.
 *
 * This module is the only place the token is held; it lives in the
 * TOKEN module variable, never in localStorage or sessionStorage, because a
 * token that outlives the run it belongs to is a liability and a fresh one
 * is generated every server start anyway.
 *
 * store.setCallouts already has one caller, pins.js's refetchCallouts. This
 * module never calls it directly and never writes state itself; it only
 * decides, via startCalloutsPoll/stopCalloutsPoll/refetchCallouts handed in
 * by main.js, whether the poll or the push is the one presently allowed to
 * call it, so the two are never both running at once.
 *
 * `onHello` and `onAgentEvent` are the chat pane's two inbound hooks:
 * `onHello` receives the hello frame's `session` object once per connection
 * (including every reconnect, since agent status can change between them),
 * and `onAgentEvent` receives every event whose kind is not
 * `callouts_changed`. Neither is called from here except at those two
 * points; chat.js, not this module, decides what an event means. The
 * returned `send` is this module's only outbound capability, so a turn,
 * permission or interrupt frame still goes out over the one socket this
 * closure owns, with no second connection or second reconnect policy.
 */

import { store } from "./store.js";
import { showError, toast } from "./ui.js";

const PROTOCOL_VERSION = 1;

const BASE_BACKOFF_MS = 500; // also the "one backoff interval" the fallback poll waits out
const MAX_BACKOFF_MS = 15000;

// How long a connection attempt is given to prove it is alive, from the
// moment connect() constructs the socket: past this, with no message
// (open's own hello included) having reset the clock, the socket is
// treated as dead whether it never reached OPEN at all or opened and then
// went quiet with no close event of its own (a laptop that slept, a NAT
// mapping dropped silently).
//
// This must stay at least twice the server's PING_INTERVAL (http/ws.py,
// currently 5 seconds), because those pings are the only traffic an
// otherwise idle connection carries and they are what distinguishes idle
// from dead. Set below that and a healthy connection with nothing to say
// gets closed and reopened forever; set far above it and a phone that lost
// its network waits too long before falling back to polling. Three missed
// pings is the compromise.
const LIVENESS_TIMEOUT_MS = 15000;

const REFUSED_MESSAGE =
  "This page's link has gone stale (the server was restarted, or this " +
  "tab was opened without the printed link, since a plain reload does not " +
  "keep the token). Reopen the URL printed in the terminal to get a " +
  "working one.";
const MISMATCH_MESSAGE =
  "This page is running an older build than the server now speaks. " +
  "Reload the page to pick up the current one.";

const CONN_LABEL = {
  connecting: "Connecting…",
  live: "Live",
  polling: "Polling",
  refused: "Reopen URL",
};
const CONN_TITLE = {
  connecting: "Connecting to the live update channel.",
  live: "Live: callouts update without a reload.",
  polling: "Live updates are unavailable right now; falling back to checking for callouts every 1.5s.",
  refused: REFUSED_MESSAGE,
};

/**
 * Reads and consumes the per-run token from location.hash ("#t=<token>").
 * A fragment is never sent to a server, which is why the token travels
 * here rather than in the path or a query parameter for the initial page
 * load; it is promoted to a query parameter only for the one request that
 * needs it, the /ws handshake, because a browser cannot set headers on
 * that handshake and this is the only channel left that is not the
 * subprotocol field.
 *
 * The fragment is stripped from the visible URL unconditionally, whether
 * or not a token was found in it: a token left visible in the address bar
 * survives a reload, a bookmark and a screen share, all of which are wider
 * exposure than the "never sent to a server" property the fragment was
 * chosen for in the first place.
 */
function extractToken() {
  const hash = location.hash;
  let token = "";
  if (hash.startsWith("#")) {
    token = new URLSearchParams(hash.slice(1)).get("t") || "";
  }
  if (hash) {
    history.replaceState(null, "", location.pathname + location.search);
  }
  return token;
}

const TOKEN = extractToken();

function makeTabId() {
  if (window.crypto && typeof window.crypto.randomUUID === "function") {
    return window.crypto.randomUUID();
  }
  // Every browser this viewer targets has crypto.randomUUID; this only
  // covers an embedded WebView that stripped it. Uniqueness here only
  // needs to hold across the handful of tabs one person opens, not
  // cryptographic strength, since the tab id carries no authority: the
  // token is what proves who is allowed to connect at all.
  return "t" + Math.random().toString(36).slice(2) + Date.now().toString(36);
}

const TAB_ID = makeTabId();

function wsPath() {
  return "/ws?t=" + encodeURIComponent(TOKEN);
}

function wsUrl() {
  const scheme = location.protocol === "https:" ? "wss:" : "ws:";
  return scheme + "//" + location.host + wsPath();
}

export function initWs({
  startCalloutsPoll,
  stopCalloutsPoll,
  refetchCallouts,
  onHello = () => {},
  onAgentEvent = () => {},
}) {
  let ws = null;
  let opened = false; // true once this attempt's WebSocket has reached readyState OPEN
  let lastSeq = 0;
  let attempt = 0;
  let stopped = false; // permanently done trying: protocol mismatch or a confirmed 403
  let fallbackFired = false; // whether this downtime episode has already switched to polling
  let fallbackTimer = null;
  let reconnectTimer = null;
  let livenessTimer = null;

  const connIndicator = document.getElementById("connIndicator");
  function applyConnIndicator(state) {
    connIndicator.textContent = CONN_LABEL[state.connection] || state.connection;
    connIndicator.title = CONN_TITLE[state.connection] || "";
    connIndicator.dataset.state = state.connection;
  }
  applyConnIndicator(store.getState());
  store.subscribe("connection", applyConnIndicator);

  function backoffDelay(n) {
    const raw = Math.min(MAX_BACKOFF_MS, BASE_BACKOFF_MS * Math.pow(2, n));
    // +/-25% jitter, so a server restart does not get every tab's next
    // retry landing on it in the same instant.
    return raw * (0.75 + Math.random() * 0.5);
  }

  function send(frame) {
    if (ws && ws.readyState === WebSocket.OPEN) ws.send(JSON.stringify(frame));
  }

  function clearLiveness() {
    clearTimeout(livenessTimer);
    livenessTimer = null;
  }

  function resetLiveness() {
    clearTimeout(livenessTimer);
    livenessTimer = setTimeout(handleLivenessExpiry, LIVENESS_TIMEOUT_MS);
  }

  function handleLivenessExpiry() {
    livenessTimer = null;
    if (stopped || !ws) return;
    // Nothing reset the deadline: this attempt's socket never delivered a
    // single message since connect(), whether because it never reached
    // OPEN, opened but never got a hello back, or went quiet sometime
    // after that with no close event of its own. ws.close() is the one
    // action that produces a decision in every one of those cases: fired
    // while still CONNECTING it closes without ever having opened, which
    // handleClose routes to probeRefusal; fired on an OPEN socket it
    // routes to scheduleRetry instead.
    try {
      ws.close();
    } catch (err) {
      // already closing or closed; handleClose was already called or is imminent
    }
  }

  function connect() {
    if (stopped) return;
    clearTimeout(reconnectTimer);
    reconnectTimer = null;
    opened = false;
    try {
      ws = new WebSocket(wsUrl());
    } catch (err) {
      // Malformed URL or a browser that refuses to construct a WebSocket
      // (e.g. mixed content); nothing to learn from this beyond "try again
      // later", the same as any other pre-handshake failure.
      probeRefusal();
      return;
    }
    resetLiveness();
    ws.addEventListener("open", handleOpen);
    ws.addEventListener("message", handleMessage);
    ws.addEventListener("close", handleClose);
    // No separate handling in "error": the WebSocket spec fires close right
    // after error for every failure mode, and close is where readyState and
    // the close code are both available, so all decisions are made there.
    ws.addEventListener("error", () => {});
  }

  function handleOpen() {
    opened = true;
    send({
      v: PROTOCOL_VERSION,
      type: "hello",
      token: TOKEN,
      last_seq: lastSeq,
      viewer: { tab_id: TAB_ID, w: window.innerWidth, h: window.innerHeight },
    });
  }

  function handleMessage(event) {
    // Any inbound frame, parseable or not, proves the transport is still
    // alive; this is the one reset shared by the "never got a hello"
    // and "went idle after hello" halves of the liveness deadline.
    resetLiveness();
    let frame;
    try {
      frame = JSON.parse(event.data);
    } catch (err) {
      return; // not a frame this client understands; nothing was sent that depends on a reply
    }
    if (frame.v !== PROTOCOL_VERSION) {
      // The server is expected to close with 4400 rather than send a frame
      // with the wrong v to a client that announced the right one; this is
      // a defensive second check in case a future server build ever sends
      // one frame before closing.
      handleProtocolMismatch();
      return;
    }
    if (frame.type === "hello") {
      handleHello(frame);
    } else if (frame.type === "event") {
      handleEvent(frame);
    } else if (frame.type === "refused") {
      // Always sent in answer to something this page sent, and always
      // carrying a reason. Dropping it would leave the human watching for an
      // effect that is never coming, with the server's explanation of why
      // discarded one layer below the UI.
      toast(frame.reason || "the server refused that request", false);
    }
    // "call" and "ping": no handler yet. An unrecognised type from a
    // same-version server is simply a frame this build of the page has no
    // use for, not an error.
  }

  function handleHello(frame) {
    if (frame.protocol !== PROTOCOL_VERSION) {
      handleProtocolMismatch();
      return;
    }
    lastSeq = frame.seq;
    attempt = 0;
    fallbackFired = false;
    clearTimeout(fallbackTimer);
    fallbackTimer = null;
    store.setConnection("live");
    stopCalloutsPoll();
    // One refetch on every (re)connect, in addition to reacting to
    // callouts_changed below: replay only covers events still in the
    // server's 500-event ring, so a gap longer than that would otherwise
    // leave a stale callout list with no further event to prompt a refetch.
    refetchCallouts();
    onHello(frame.session);
  }

  function handleEvent(frame) {
    lastSeq = frame.seq;
    const kind = frame.event && frame.event.kind;
    if (kind === "callouts_changed") {
      refetchCallouts();
    } else {
      onAgentEvent(frame.event);
    }
  }

  function handleProtocolMismatch() {
    stopped = true;
    clearLiveness();
    clearTimeout(fallbackTimer);
    clearTimeout(reconnectTimer);
    if (ws) {
      try {
        ws.close();
      } catch (err) {
        // already closing or closed; nothing left to do
      }
    }
    showError(MISMATCH_MESSAGE);
    store.setConnection("polling");
    startCalloutsPoll();
  }

  function handleClose(event) {
    const wasOpened = opened;
    opened = false;
    ws = null;
    clearLiveness();
    if (stopped) return;
    if (wasOpened && event.code === 4400) {
      handleProtocolMismatch();
      return;
    }
    if (wasOpened) {
      // A drop after a successful handshake: the token, Origin and Host
      // checks already passed once for this page, so there is nothing to
      // diagnose, only a network blip, a laptop sleep or a server restart
      // to wait out.
      //
      // The state has to stop saying "live" here, before scheduleRetry arms
      // the fallback timer, because that timer's own guard skips the switch
      // to polling when it finds the connection live. Leaving it live across
      // a drop makes that guard read a state this close never cleared, and
      // the fallback never fires at all: the page then shows a live socket
      // and runs neither the push nor the poll, which is the exact outcome
      // the fallback exists to prevent.
      store.setConnection("connecting");
      scheduleRetry();
      return;
    }
    // Never reached "open". Per the WebSocket platform contract this looks
    // identical whether the cause was a wrong token, a rejected Origin or
    // Host, or the server simply being unreachable: every browser reports
    // a pre-handshake failure as the same reasonless abnormal closure, with
    // no status code and no body exposed to script. probeRefusal tells
    // a stale token apart from a transient outage by asking the same
    // question over plain HTTP instead, where the status code is visible.
    probeRefusal();
  }

  async function probeRefusal() {
    if (stopped) return;
    try {
      const res = await fetch(wsPath(), { cache: "no-store" });
      if (res.status === 403) {
        // Confirmed: this exact request, over a channel that does expose
        // its status, was refused before any upgrade was attempted. Per
        // the auth design, 403 on /ws is emitted by the token, Origin and
        // Host checks, which is the only way a request already reaching
        // this same origin fails here; further retries with the same
        // fixed token would repeat forever, so the socket path is
        // abandoned for this page load and the fallback poll takes over
        // for good.
        //
        // Only 403 is terminal, and every other status falls through to a
        // retry on purpose. A healthy server answers this same probe with
        // 400, because the probe is a plain GET carrying no upgrade headers
        // and the route reaches the handshake and rejects it there: that
        // status means the token was accepted and the socket failure was
        // transient, which is exactly the case worth retrying.
        stopped = true;
        clearTimeout(fallbackTimer);
        fallbackTimer = null;
        store.setConnection("refused");
        showError(REFUSED_MESSAGE);
        startCalloutsPoll();
        return;
      }
    } catch (err) {
      // The probe failed too (offline, DNS, TLS): no more information than
      // the WebSocket attempt already gave. Falls through to the ordinary
      // retry path below.
    }
    scheduleRetry();
  }

  function scheduleRetry() {
    if (stopped) return;
    if (!fallbackFired && fallbackTimer === null) {
      // Only armed once per downtime episode: repeated failures before it
      // fires must not stack a second timer that fires again later and
      // re-runs the switch-to-polling step redundantly.
      fallbackTimer = setTimeout(() => {
        fallbackTimer = null;
        if (stopped || fallbackFired || store.getState().connection === "live") return;
        fallbackFired = true;
        store.setConnection("polling");
        startCalloutsPoll();
      }, BASE_BACKOFF_MS);
    }
    reconnectTimer = setTimeout(connect, backoffDelay(attempt));
    attempt++;
  }

  connect();

  // `send` is the only capability this module hands back: chat.js's turn,
  // permission and interrupt frames all go out through it, so there is
  // still exactly one WebSocket and one place (this closure) that owns its
  // lifecycle, reconnects and replay bookkeeping.
  return { send };
}

File: src/annealage_mesh/http/ws.py
"""The ``/ws`` route: authentication, the greeting, replay, and frame dispatch.

A bug in this file is not a UI bug. ``/ws`` is the one route a browser tab
turns into a live bidirectional channel, and from the milestone that lands an
agent it carries turns into a Claude Code session with shell access in the
human's project. A check that can be bypassed here is a remote-code-execution
bug, so the three checks below are independent, each is commented with what it
defends against and why the other two do not cover it, and all three run
**before** the WebSocket handshake.

Refusing before the handshake, rather than upgrading and then closing with a
code, is deliberate and load-bearing in two ways. It means no connection
object, no queue and no writer task is ever created for a request that failed
authentication, so a refused caller cannot consume per-connection resources.
And it is the only form of refusal that microdot's in-process test client can
observe at all: that client's fake socket discards every outbound frame whose
opcode is not TEXT or BINARY, so a test asserting a close code through it
asserts on a frame the harness never delivered and cannot fail. An HTTP status
is checkable; a close code, on that path, is not.

Every refusal returns the same status and the same body. A response that
distinguished "no token" from "wrong token", or "bad Origin" from "bad Host",
would answer questions an unauthenticated caller should not be able to ask.
"""

import asyncio
import json
import sys
import time

from microdot import Response
from microdot.websocket import WebSocket, WebSocketError, websocket_upgrade

from .. import protocol
from ..session.base import UnknownRequest
from ..viewers import ViewerRegistry

# Inbound frame ceiling. Set explicitly because microdot's default,
# ``max_message_length = -1``, means "fall back to Request.max_body_length",
# which app.py raises to 8 MiB for pin submissions: without this line, an
# unrelated route's upload allowance silently becomes the per-frame buffer
# ceiling on every socket, and a frame is buffered whole before it is
# validated. 256 KiB comfortably holds the largest frame the protocol
# defines, a turn of text plus a permission decision; images arrive over
# POST /upload rather than over the socket, so nothing legitimate here is
# anywhere near this size.
MAX_WS_MESSAGE = 256 * 1024

# How often every connected viewer is pinged.
#
# This is what makes the browser's own liveness watchdog meaningful, and the
# two numbers are a pair: static/js/ws.js closes a socket that has delivered
# nothing for LIVENESS_TIMEOUT_MS, so without a ping this interval, an idle but
# perfectly healthy connection looks dead and gets closed and reopened on a
# loop. The watchdog there must stay at least twice this value, and says so.
#
# A dead-but-open socket is the case that needs it: a phone that slept or a
# tailnet that rekeyed leaves a connection whose readyState is still OPEN for
# minutes, with no close event, so the only way the page learns is by noticing
# that nothing is arriving any more.
PING_INTERVAL = 5.0
_REFUSED_BODY = "forbidden"
_REFUSED_STATUS = 403


def refusal():
    """The one response every failed authentication check returns."""
    return Response(_REFUSED_BODY, _REFUSED_STATUS)


def host_is_allowed(req, allowed_hosts):
    """True if this request's ``Host`` header may name this server.

    An absent ``Host`` is allowed. HTTP/1.1 requires the header and every
    browser sends it, so absent means a non-browser client (curl, a test
    harness, a script), which is not the caller this check exists for. The
    caller it exists for is a browser that has been pointed at this server by
    a hostname whose DNS answer an attacker controls and has repointed at this
    address after the page loaded. That page's ``Origin`` is the attacker's
    own site, so the Origin check catches the browser case; this check catches
    the request that carries no Origin at all, and it applies to every route
    rather than only ``/ws``, because a rebound name can read ``/manifest``
    and write through ``/submit`` just as readily as it can open a socket.

    Comparison is against the literal set of spellings that can truthfully
    name this bind, with no parsing of the header. A parser would have to
    decide what to do with userinfo, a trailing dot, and unbalanced IPv6
    brackets, and each of those decisions is a chance to extract an address
    from a header that does not actually name it.
    """
    host = req.headers.get("Host")
    if host is None:
        return True
    return host in allowed_hosts


def register_ws(app, *, token, allowed_origins, allowed_hosts, registry,
                event_log, session_info, session=None):
    """Register ``/ws`` on ``app``.

    ``token`` of ``None`` means no token was configured for this process, and
    every ``/ws`` request is then refused. That is the safe reading rather
    than an inconvenient one: a socket nobody can authenticate is a socket
    that should not open, and the command-line entry point always generates a
    token, so the only way to reach this state is a caller that built an app
    without one.
    """

    @app.get("/ws")
    async def ws_route(req):
        # Order is token, then Origin. Host is enforced for every route by
        # app.py's before_request hook and so has already passed by here.
        if not _token_is_allowed(req, token):
            return refusal()
        if not _origin_is_allowed(req, allowed_origins):
            return refusal()

        ws = await websocket_upgrade(req)
        ws.max_message_length = MAX_WS_MESSAGE
        conn = None
        try:
            # Greeting and replay both happen before the connection is
            # registered, and that order is load-bearing. Registering starts a
            # writer task that also sends on this socket, and a replayed event
            # carries an older seq than anything live: if a replay frame were
            # written from here while that writer was draining a live
            # broadcast, the client could see an older seq after a newer one,
            # move its resync position backwards, and replay the same events
            # again on its next reconnect. With nothing registered yet, there
            # is no second writer for those frames to race.
            await ws.send(json.dumps(protocol.build_hello(
                event_log.current_seq,
                session_info["id"],
                session_info.get("sdk_session_id"),
                session_info["cwd"],
                session_info.get("agent", "unavailable"),
            )))
            if not await _greet(ws, event_log, token):
                return Response.already_handled
            conn = await registry.add(ws)
            await _serve_connection(ws, conn, registry, event_log, token, session)
        except WebSocketError:
            # The peer closed, or sent a frame microdot could not read. Not
            # an error worth reporting: a browser tab closing is the ordinary
            # end of every connection.
            pass
        finally:
            if conn is not None:
                await registry.remove(conn)
        return Response.already_handled


def _token_is_allowed(req, token):
    """Constant-time check of the ``t`` query parameter against the run's token.

    The token travels as a query parameter for exactly one reason: a browser
    cannot set headers on a WebSocket handshake, so a header is not available
    on the only route that needs this. It stays out of our own stderr because
    the access log prints ``req.path``, which microdot splits from the query
    string, and never ``req.query_string``.

    A missing token is compared against the configured one anyway rather than
    returning early, so the refusal path costs the same either way and a
    missing token cannot be told apart from a wrong one.

    A request supplying ``t`` more than once is refused outright rather than
    resolved. ``req.args`` is a MultiDict whose lookup returns the first value,
    so ``?t=<real>&t=wrong`` would otherwise authenticate: that is fine on its
    own, since a caller who already knows the token gains nothing, but it means
    the answer to "which value counts" is a property of one parser. Anything in
    front of this server, a reverse proxy or ``tailscale serve``, may pick the
    other one, and a check whose result depends on which component looked is a
    check waiting to disagree with itself. One value, or no.
    """
    if not token:
        return False
    supplied_values = req.args.getlist("t") if hasattr(req.args, "getlist") \
        else [req.args["t"]] if "t" in req.args else []
    if len(supplied_values) > 1:
        return False
    supplied = supplied_values[0] if supplied_values else ""
    return _constant_time_equal(supplied, token)


def _constant_time_equal(a, b):
    import hmac
    return hmac.compare_digest(a.encode("utf-8", "surrogatepass"),
                               b.encode("utf-8", "surrogatepass"))


def _origin_is_allowed(req, allowed_origins):
    """True if this request's ``Origin`` is one this bind serves.

    An absent ``Origin`` is allowed, and the reason is specific: a browser
    always sends one on a WebSocket handshake, so absent means a non-browser
    client, which still has to present the token. The threat this check exists
    for is a page the human happens to visit opening a socket to this server,
    and that page cannot suppress its own ``Origin``.

    This is not redundant with the token. **A WebSocket handshake is not
    subject to the same-origin policy**, so a malicious page can open this
    socket cross-origin and read every reply; without this check, the only
    thing between such a page and an agent with shell access would be whether
    the token ever appeared somewhere a page could read.

    Membership is exact. Prefix or substring matching would accept
    ``http://127.0.0.1`` (a truncation) and
    ``http://evil.example/http://127.0.0.1:8765`` (a containment), neither of
    which is a value this server ever issued.
    """
    origin = req.headers.get("Origin")
    if origin is None:
        return True
    return origin in allowed_origins


async def _greet(ws, event_log, token):
    """Read the client's ``hello`` and answer it, before any writer exists.

    Returns False if the connection was closed here, in which case the caller
    must not register it. Nothing else is read: a client that sends some other
    frame first is answered with a refusal and asked again, because ``hello``
    is what carries the resync position and there is nothing useful to do
    without it.

    A client that connects and then says nothing is never registered, so it
    receives no events and consumes no writer task. That is the right outcome
    rather than an oversight: it is not a viewer until it identifies itself,
    and replay from its ``last_seq`` covers everything it misses in the gap.
    """
    while True:
        raw = await ws.receive()
        frame = await _parse(ws, raw)
        if frame is None:
            continue
        if frame is _CLOSED:
            return False
        if frame["type"] != "hello":
            await ws.send(json.dumps(protocol.build_refused(
                "expected a hello frame first, got %s" % frame["type"])))
            continue
        # The token is re-checked here even though the query parameter already
        # authenticated the connection. It costs one comparison, and a hello
        # whose token disagrees with the one that opened the socket is a
        # confused or hostile client either way.
        if not _constant_time_equal(frame.get("token") or "", token or ""):
            await protocol.close_with_code(
                ws, protocol.CLOSE_VERSION_MISMATCH, "hello token does not match")
            return False
        replay = event_log.replay(frame.get("last_seq"))
        for seq, event_wire in replay.events:
            await ws.send(json.dumps(protocol.build_event(seq, event_wire)))
        if replay.truncated:
            await ws.send(json.dumps(protocol.build_refused(
                "history before this point is not in the live ring; page it over HTTP")))
        return True


#: Returned by _parse when it has already closed the connection.
_CLOSED = object()


async def _parse(ws, raw):
    """Parse and validate one inbound frame.

    Returns the frame, None if it was refused and the connection continues, or
    ``_CLOSED`` if the version mismatch ended it. A protocol-version mismatch
    is the one failure that ends a connection with a coded close rather than an
    HTTP status, because by definition it is discovered after the handshake:
    the version is inside a frame. 4400 is what separates a stale cached page
    from a newer server.
    """
    try:
        parsed = json.loads(raw)
    except (ValueError, TypeError):
        await ws.send(json.dumps(protocol.build_refused("frame is not valid JSON")))
        return None
    try:
        ok, result = protocol.validate_inbound(parsed)
    except protocol.ProtocolVersionMismatch as mismatch:
        await protocol.close_with_code(
            ws, protocol.CLOSE_VERSION_MISMATCH,
            "protocol version %s, this server speaks %d"
            % (mismatch.args[0] if mismatch.args else "?", protocol.PROTOCOL_VERSION))
        return _CLOSED
    if not ok:
        await ws.send(json.dumps(protocol.build_refused(result)))
        return None
    return result


async def _serve_connection(ws, conn, registry, event_log, token, session=None):
    """Read and dispatch frames until the peer goes away.

    Every frame this sends is either a ``refused``, which carries no seq, or a
    terminal coded close. Neither has an ordering relationship with the events
    the connection's writer task is draining onto the same socket, so writing
    them from here cannot reorder anything the client tracks. Frames cannot
    interleave mid-frame either: microdot's ``awrite`` is a single
    ``StreamWriter.write`` of the whole frame followed by a ``drain``, and
    ``write`` buffers synchronously.
    """
    while True:
        raw = await ws.receive()
        frame = await _parse(ws, raw)
        if frame is None:
            continue
        if frame is _CLOSED:
            return
        await _dispatch(ws, conn, registry, event_log, token, frame, session)


async def _dispatch(ws, conn, registry, event_log, token, frame, session=None):
    """Route one validated inbound frame."""
    kind = frame["type"]
    if kind == "hello":
        # The opening hello is answered by _greet, before this connection was
        # registered. A second one is not a resync request: replaying from here
        # would write older seqs onto a socket whose writer task is draining
        # newer ones. It is treated as interaction, which re-elects this viewer
        # as the primary, and refused with that said plainly.
        await registry.touch(conn)
        await ws.send(json.dumps(protocol.build_refused(
            "hello was already answered; reconnect to resync from a last_seq")))
        return
    if kind in ("result", "error"):
        registry.resolve_call(conn, frame)
        return
    if kind == "state":
        await registry.touch(conn)
        return
    if kind in ("turn", "interrupt", "permission"):
        if session is None:
            # Viewer-only: the frames are still defined and validated so one
            # browser build works against both modes, and answering with a
            # reason is what stops a chat pane waiting forever on a turn
            # nothing will ever process.
            await ws.send(json.dumps(protocol.build_refused(
                "this server is running viewer-only, so %s frames are not served"
                % kind)))
            return
        await registry.touch(conn)
        # Handed to the session and not awaited for a result: a turn produces
        # events, which arrive over this same socket through the registry, and
        # blocking this connection's read loop on the whole turn would stop it
        # reading the interrupt frame that ends it.
        try:
            if kind == "turn":
                await session.submit_turn(frame["blocks"], viewer=conn.tab_id)
            elif kind == "interrupt":
                await session.interrupt()
            else:
                await session.decide_permission(
                    frame["request_id"], frame["decision"], frame.get("message", ""))
        except UnknownRequest:
            # Ordinary, not a failure: two tabs held one card and this is the
            # one that lost, or the request expired before the click landed.
            # Answered with what actually happened rather than with the generic
            # message below, because "already decided" tells the human their
            # click changed nothing, which is the whole point of replying.
            await ws.send(json.dumps(protocol.build_refused(
                "that permission request was already decided, by another view or "
                "by expiring; this decision was not applied")))
        except Exception as exc:
            # A session that fails must not take the socket with it: the viewer
            # half of this page keeps working whatever the agent does.
            sys.stderr.write("warning: session could not handle a %s frame: %r\n"
                             % (kind, exc))
            await ws.send(json.dumps(protocol.build_refused(
                "the agent could not handle that %s frame; see the server output" % kind)))
        return
    await ws.send(json.dumps(protocol.build_refused("unhandled frame type: %s" % kind)))


def build_registry(**kwargs):
    """A ``ViewerRegistry`` for one app, kept here so app.py does not need to
    know the registry's constructor keywords."""
    return ViewerRegistry(**kwargs)


def ping_frame():
    """A liveness frame, addressed with the wall-clock time the plan's frame
    catalogue shows."""
    return protocol.build_ping(int(time.time()))


async def ping_forever(registry, interval=PING_INTERVAL):
    """Broadcast a ping to every connected viewer, forever.

    One task for the whole process rather than one per connection: a ping
    carries no per-connection state, and the registry's backpressure policy
    already drops queued pings first, so a saturated connection sheds these
    before it sheds anything that matters.

    Without this the browser's liveness watchdog has nothing to distinguish an
    idle connection from a dead one, and closes both.
    """
    while True:
        await asyncio.sleep(interval)
        await registry.broadcast(ping_frame())


__all__ = [
    "MAX_WS_MESSAGE", "PING_INTERVAL", "WebSocket", "build_registry",
    "host_is_allowed", "ping_forever", "ping_frame", "refusal", "register_ws",
]
